### ISSUE TITLE
Stabilise visualisation module

### ISSUE THREAD

**Issue by nzw0301** [optuna/optuna#2959]
## Motivation

Optuna has two visualisation backends: Plotly and Matplotlib. It would be great to minimise the difference between backends to provide consistent plots. Concretely, 

1. https://github.com/optuna/optuna/issues/3226
2. Unify appearance between plotly and matplotlib
3. https://github.com/optuna/optuna/issues/3225
4. handle with the outliner such as `nan`/`inf`.

In the same backend, we can improve more consistency among different plot functions. Concretely, 

5. Unify appearance among plot functions

### 1. Add more Matplotlib tests

See https://github.com/optuna/optuna/issues/3226

### 2. Unify appearance between Plotly and Matplotlib

I've already created related issues:

- https://github.com/optuna/optuna/issues/2893
- https://github.com/optuna/optuna/issues/2913
- https://github.com/optuna/optuna/issues/3197

But feel free to let me know the thing that should be addressed by this issue.

### 3. Add unimplemented plot functions to `optuna.visualization.matplotlib`

- https://github.com/optuna/optuna/issues/3225

### 4. handle with the outliner such as `nan`/`inf`.

- https://github.com/optuna/optuna/issues/3382

### 5. Unify appearance among plot functions in the same backend

- https://github.com/optuna/optuna/issues/3375

**Comment by HideakiImamura** [comment]
@nzw0301 All enumerated items appear to be complete. Is it OK to close this issue?

**Comment by nzw0301** [comment]
Sorry, not yet; we need to improve the matplotlib's tests. I only added file names that had already been resolved to the issue description above.

**Comment by nzw0301** [comment]
I'll revise the issue description later. Thank you for kindly letting me know this!

**Comment by nzw0301** [comment]
We might be able to add a unit test using objective values. For example, we access `plot_optimization_history`'s data as follows:

```python
import optuna

s = optuna.create_study()

def objective(t):
    
    return t.suggest_float("x", 0., 10.)

s.optimize(objective, n_trials=20)
fig = optuna.visualization.matplotlib.plot_optimization_history(s)
print(a.get_lines()[0].get_data())  # access best value
a.collections[0].get_offsets().data  # objective value
```

**Comment by akawashiro** [comment]
I'm working on test_pareto_front.py.

**Comment by belldandyxtq** [comment]
I will be working on `test_param_importances.py`

**Comment by not522** [comment]
I found that the axes of `optuna.visualization.matplotlib.plot_contour` has bug.

```python
import optuna
from optuna.testing.visualization import prepare_study_with_trials
import matplotlib.pyplot as plt

study = prepare_study_with_trials(more_than_three=True, n_objectives=2, with_c_d=True)
print(study.trials_dataframe())
figure = optuna.visualization.plot_contour(study, target=lambda t: t.values[0])

figure.write_image('plotly.png')
optuna.visualization.matplotlib.plot_contour(study, target=lambda t: t.values[0])
plt.savefig('matplotlib.png')
```

![plotly](https://user-images.githubusercontent.com/1255726/146722151-9516e355-f9a7-4ccc-95fd-3158e5ea500b.png)

![matplotlib](https://user-images.githubusercontent.com/1255726/146722133-f7a48ce3-1985-4537-b672-bd4150e5307c.png)

To plot a-b subplot, the max and the min of params are determined with trials whose params have `a` and `b`. But the maximum and minimum values should be determined independently for `a` and `b`.

**Comment by nzw0301** [comment]
This issue looks too large now, so I'll split this PR into several sub-issues...

**Comment by HideakiImamura** [comment]
Let me add another TODO: unify and refactor the tests of plotly and matplotlib.

**Comment by HideakiImamura** [comment]
Let me close this issue since we have completed all of these items and released Optuna v3.0!!! Thank you all for your contributions!!!
